src/authorize.ts

```diff
@@ -0,0 +1,152 @@
+   * can be saved to reduce a lookup roundtrip on subsequent requests.
+   */
+  identityId: string;
+
+  /** The userId supplied by the mapToUserId function for this request. */
+  userId: string;
+}
+
+export interface IEveryAuthOptions {
+  /**
+   * The URL that the user should be sent to after the operation is complete. Can be either
+   * '/path/component/only' or a fully qualified URL.
+   *
+   * Supplied with an `error` query parameter if the operation resulted in an error.
+   */
+  finishedUrl: string;
+
+  /**
+   * Either a string or a function which generates a string that contains the protocol, hostname, port, and
+   * path to the current mount point.  If this is not supplied, the EveryAuth system attempts to determine it
+   * programmatically by looking at various attributes on the request object.
+   */
+  hostedBaseUrl?: string | ((req: express.Request) => string);
+
+  /**
+   * Map the current request to a unique userId tag that can be used to retrieve the authorized token on
+   * subsequent requests.  Usually extracts the value from the output of the authorization system, or can be a
+   * constant for experimentation and testing purposes.
+   */
+  mapToUserId: ((req: express.Request) => Promise<string>) | ((req: express.Request) => string);
+
+  /**
+   * Called after the authorization process completed.
+   *
+   * Return false to indicate that the request was responded to, and that EveryAuth should not redirect the
+   * caller to the finishedUrl automatically.
+   */
+  onComplete?: (req: express.Request, res: express.Response, everyCtx: IEveryAuthContext) => Promise<boolean>;
+}
+
+/**
+ * @ignore
+ *
+ * Tries to determine the publically-visible URL for this part of the router.
+ */
+export const getHostedBaseUrl = (options: IEveryAuthOptions, req: express.Request): string => {
+  // Supplied in options; use that.
+  if (options.hostedBaseUrl) {
+    if (typeof options.hostedBaseUrl == 'string') {
+      return options.hostedBaseUrl;
+    } else if (typeof options.hostedBaseUrl == 'function') {
+      return options.hostedBaseUrl(req);
+    }
+  }
+
+  // Discovery time!
+  //
+  // Sometimes the originalUrl includes the entire request, sometimes it doesn't!
+  if (req.originalUrl.startsWith('http')) {
+    return req.originalUrl;
+  }
+
+  // req.hostname doesn't preserve the port, unfortunately, so test the Host header to see if it's present.
+  let port = '';
+  if (req.headers['host'] && req.headers['host'].includes(':')) {
+    port = `:${req.headers['host'].split(':')[1]}`;
+  }
+
+  // Return a hopefully valid URL.
+  return `${req.protocol}://${req.hostname}${port}${req.originalUrl}`;
+};
+
+/**
+ * Authorize a user with a particular service.
+ * @param serviceId The name of the service to authorize the user against.
+ * @param options Configuration options to control the behavior of this authorization.
+ * @return An Express Router
```

Comment:
Actually, this function is not returning an Express Router,  it performs a redirect, should be void.
Add that to the JSDoc description.